### PR TITLE
BUG: update should try harder to preserve dtypes #4094

### LINKED ISSUE #4094: BUG: update should try harder to preserve dtypes
more examples in #13957 

http://stackoverflow.com/questions/17398216/unwanted-type-conversion-in-pandas-dataframe-update

```
df = pd.DataFrame({'int': [1, 2], 'float': [np.nan, np.nan]})

print('Integer column:')
print(df['int'])

for _, df_sub in df.groupby('int'):
    df_sub['float'] = float(df_sub['int'])
    df.update(df_sub)

print('NO integer column:')
print(df['int']) 
```

### PR DESCRIPTION
**Pull request by Horstage**
- [ ] closes #4094
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [x] whatsnew entry

I have to admit, I am not quite sure about the code in the [initial example](https://github.com/pandas-dev/pandas/issues/4094#issue-16212879). But the rest of the given examples ([1](https://github.com/pandas-dev/pandas/issues/4094#issuecomment-297671925), [2](https://github.com/pandas-dev/pandas/issues/4094#issuecomment-583174514)) are converted to tests and they do pass.

I would have loved to make better use of hypothesis in the tests, but currently [column](https://hypothesis.readthedocs.io/en/latest/numpy.html#hypothesis.extra.pandas.columns) does not support dtype strategies and neither does [from_dtype](https://hypothesis.readthedocs.io/en/latest/numpy.html#hypothesis.extra.numpy.from_dtype) (if I would define the elements of the column and infere the dtype).

### CODE REVIEW

File: pandas/tests/frame/methods/test_update.py [hunk 1]
```diff
@@ -146,3 +148,43 @@ def test_update_with_different_dtype(self):
 
         expected = DataFrame({"a": [1, 3], "b": [np.nan, 2], "c": ["foo", np.nan]})
         tm.assert_frame_equal(df, expected)
+
+    @given(dtype=np_st.integer_dtypes(endianness="="))
```
Comment (gfyoung) [inline pandas/tests/frame/methods/test_update.py:152]: Why is this needed?

Comment (Horstage) [reply pandas/tests/frame/methods/test_update.py:152]: If you talk about endianness: Locally I there was an
> ValueError: Big-endian buffer not supported on little-endian compiler

so I used only native endian.

```
Falsifying example: test_update_with_subset_and_same_integer_dtype(
    self=<pandas.tests.frame.methods.test_update.TestDataFrameUpdate at 0x7f12171ac940>,
    dtype=dtype('>i2'),
)
Traceback (most recent call last):
  File "/home/jhuennemeyer/projects/pandas/pandas/tests/frame/methods/test_update.py", line 157, in test_update_with_subset_and_same_integer_dtype
    df.update(update)
  File "/home/jhuennemeyer/projects/pandas/pandas/core/frame.py", line 6947, in update
    other = other.reindex_like(self)
  File "/home/jhuennemeyer/projects/pandas/pandas/core/generic.py", line 4149, in reindex_like
    return self.reindex(**d)
  File "/home/jhuennemeyer/projects/pandas/pandas/util/_decorators.py", line 323, in wrapper
    return func(*args, **kwargs)
  File "/home/jhuennemeyer/projects/pandas/pandas/core/frame.py", line 4499, in reindex
    return super().reindex(**kwargs)
  File "/home/jhuennemeyer/projects/pandas/pandas/core/generic.py", line 4839, in reindex
    return self._reindex_multi(axes, copy, fill_value)
  File "/home/jhuennemeyer/projects/pandas/pandas/core/frame.py", line 4405, in _reindex_multi
    return self._reindex_with_indexers(
  File "/home/jhuennemeyer/projects/pandas/pandas/core/generic.py", line 4908, in _reindex_with_indexers
    new_data = new_data.reindex_indexer(
  File "/home/jhuennemeyer/projects/pandas/pandas/core/internals/managers.py", line 1305, in reindex_indexer
    new_blocks = [
  File "/home/jhuennemeyer/projects/pandas/pandas/core/internals/managers.py", line 1306, in <listcomp>
    blk.take_nd(
  File "/home/jhuennemeyer/projects/pandas/pandas/core/internals/blocks.py", line 1300, in take_nd
    new_values = algos.take_nd(
  File "/home/jhuennemeyer/projects/pandas/pandas/core/array_algos/take.py", line 72, in take_nd
    return _take_nd_ndarray(arr, indexer, axis, out, fill_value, allow_fill)
  File "/home/jhuennemeyer/projects/pandas/pandas/core/array_algos/take.py", line 116, in _take_nd_ndarray
    func(arr, indexer, out, fill_value)
  File "pandas/_libs/algos_take_helper.pxi", line 1181, in pandas._libs.algos.take_2d_axis0_int16_float64
ValueError: Big-endian buffer not supported on little-endian compiler
```


If you talk about the complete line: Test parameterization. Like I said, I would like to do that differently, but I'm limited but by hypothesis.

Comment (gfyoung) [reply pandas/tests/frame/methods/test_update.py:152]: For CI purposes, can you push a commit that removes the `given` decorator.

I know it will cause problems locally, but it would be good see what the CI says.

Comment (Horstage) [reply pandas/tests/frame/methods/test_update.py:152]: Let me get this straight.
When you say:
> For CI purposes, can you push a commit that removes the `given` decorator.

I understand:
>Remove the complete line.

This line has its' purpose.
```
@given(dtype=np_st.integer_dtypes(endianness="="))
```
is the same as
```
@pytest.mark.parametrize("dtype", [np.dtype("=i1"), np.dtype("=i2"), np.dtype("=i4"), np.dtype("=i8")])
```
only that I don't have to worry about the exact parameters. That's what hypothesis does.
So removing it, just forces me to choose a specific dtype in the test. Probably I would simply use `int` and the error above is gone.
If you are still unclear about test parametrization or the hypothesis package let me know.
You can also check out [hypothesis](https://hypothesis.readthedocs.io/en/latest/quickstart.html) or [pytests docs  on that topic](https://docs.pytest.org/en/stable/fixture.html#fixture-parametrize).

By using hypothesis, I might have uncovered a very different bug.
The error, I got  locally, only arises if I choose any of these dtypes `[np.dtype(">i2"), np.dtype(">i4"), np.dtype(">i8")]` as
```
@given(dtype=np_st.integer_dtypes())
```
translates to
```
@pytest.mark.parametrize("dtype", [np.dtype("=i1"), np.dtype("=i2"), np.dtype("=i4"), np.dtype("=i8"), np.dtype(">i1"), np.dtype(">i2"), np.dtype(">i4"), np.dtype(">i8"), np.dtype("<i1"), np.dtype("<i2"), np.dtype("<i4"), np.dtype("<i8")])
```

So I will check out what your CI pipeline thinks about that.

File: doc/source/whatsnew/v1.2.4.rst [hunk 2]
```diff
@@ -25,7 +25,7 @@ Fixed regressions
 Bug fixes
 ~~~~~~~~~
 
--
+- Bug in :meth:`~DataFrame.update` unnecessarily changing the dtype of a column (:issue:`4094`)
```
Comment (jreback) [inline doc/source/whatsnew/v1.2.4.rst:28]: move to 1.3 bug fixes conversion

File: pandas/core/frame.py [hunk 3]
```diff
@@ -6939,6 +6939,7 @@ def update(
         if not isinstance(other, DataFrame):
             other = DataFrame(other)
 
+        other_dtypes = other.dtypes  # dtype might change during reindexing
```
Comment (jreback) [inline pandas/core/frame.py:6942]: don't need the comment

File: pandas/core/frame.py [hunk 4]
```diff
@@ -6963,7 +6964,11 @@ def update(
             if mask.all():
                 continue
 
-            self[col] = expressions.where(mask, this, that)
+            col_array = expressions.where(mask, this, that)
```
Comment (jreback) [inline pandas/core/frame.py:6967]: what are you trying to do here? infer and cast? this is very awkward what you are doing.
.where casts to the most appropriate type, but then you are casting back?

File: pandas/tests/frame/methods/test_update.py [hunk 5]
```diff
@@ -146,3 +148,43 @@ def test_update_with_different_dtype(self):
 
         expected = DataFrame({"a": [1, 3], "b": [np.nan, 2], "c": ["foo", np.nan]})
         tm.assert_frame_equal(df, expected)
+
+    @given(dtype=np_st.integer_dtypes())
```
Comment (jreback) [inline pandas/tests/frame/methods/test_update.py:152]: don't use hypthoesis, its if for very specific tests. we *already* have a fixture for this. any_nullable_dtype (there are many fixtures if you want to comprehensively test this in pandas/conftest.py)

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/frame/methods/test_update.py (hunk 5 above)
Comment (Horstage) [reply pandas/tests/frame/methods/test_update.py:152]: Alright! Thanks for the hint, I'll look it up.

File: pandas/core/frame.py (hunk 4 above)
Comment (Horstage) [reply pandas/core/frame.py:6967]: `.reindex_like` like changes the dtype to a nullable dtype if `other.index` is a subset of `self.index`.
`.where` receives a nullable dtype and continues to use it, even though the NaNs are masked.
I catch the case where self[col] and other[col] originally have the same dtype, but `.where` returns a different dtype.
With the assumption, that all NaNs are masked I am casting back, yes.

Example:
```
self[col] = Series(["a", "b"], dtype=str)
other[col] = Series(["c"], dtype=str)
.reindex -> other[col] = Series(["c", None], dtype=object)
.where -> self[col] = Series(["c", "b"], dtype=object)
```

File: pandas/tests/frame/methods/test_update.py (hunk 1 above)
Comment (Horstage) [reply pandas/tests/frame/methods/test_update.py]: Resolved with [this comment](https://github.com/pandas-dev/pandas/pull/40219#discussion_r587485685).

File: pandas/core/frame.py (hunk 4 above)
Comment (jreback) [reply pandas/core/frame.py:6967]: where is `.reindex_like` called? this should be fixed at the sort not patched over

**Comment by simonjayhawkins** [comment]
@Horstage closing as stale. ping when ready to continue

File: pandas/core/frame.py (hunk 4 above)
Comment (guziy) [reply pandas/core/frame.py:6967]: reindex_like is called here: https://github.com/pandas-dev/pandas/pull/40219/files#diff-421998af5fe0fbc54b35773ce9b6289cae3e8ae607f81783af04ebf1fbcaf077R6943 to have the same index in the other as in self, but for missing indices it inserts NANs and changes data type, maybe reindex_like needs to be changed in this case...
